File: internal/application/usecase/role/list.go
package role

import (
	"context"

	"go-clean-template/internal/domain"
)

type ListRolesInput struct {
	Page     int
	PageSize int
}

type ListRolesOutput struct {
	Roles      []*domain.Role
	Total      int64
	Page       int
	PageSize   int
	TotalPages int
}

type ListRolesUseCase interface {
	Execute(ctx context.Context, input *ListRolesInput) (*ListRolesOutput, error)
}

type listRolesUseCase struct {
	roleRepo domain.RoleRepository
}

func NewListRolesUseCase(roleRepo domain.RoleRepository) ListRolesUseCase {
	return &listRolesUseCase{roleRepo: roleRepo}
}

func (uc *listRolesUseCase) Execute(ctx context.Context, input *ListRolesInput) (*ListRolesOutput, error) {
	if input.Page < 1 {
		input.Page = 1
	}
	if input.PageSize < 1 {
		input.PageSize = 20
	}
	if input.PageSize > 100 {
		input.PageSize = 100
	}
	offset := (input.Page - 1) * input.PageSize
	roles, total, err := uc.roleRepo.FindAll(ctx, offset, input.PageSize)
	if err != nil {
		return nil, err
	}

	totalPages := int(total / int64(input.PageSize))
	if total%int64(input.PageSize) > 0 {
		totalPages++
	}
	if totalPages < 1 {
		totalPages = 1
	}

	return &ListRolesOutput{
		Roles:      roles,
		Total:      total,
		Page:       input.Page,
		PageSize:   input.PageSize,
		TotalPages: totalPages,
	}, nil
}

File: src/application/usecase/role/list.rs
use std::sync::Arc;

use async_trait::async_trait;

use crate::application::error::AppError;
use crate::application::usecase::user::RoleRepository;
use crate::domain;

pub struct ListRolesInput {
    pub page: i32,
    pub page_size: i32,
}

pub struct ListRolesOutput {
    pub roles: Vec<domain::Role>,
    pub total: i64,
    pub page: i32,
    pub page_size: i32,
}

#[async_trait]
pub trait ListRolesUseCase: Send + Sync {
    async fn execute(&self, input: ListRolesInput) -> Result<ListRolesOutput, AppError>;
}

pub struct ListRolesUseCaseImpl {
    role_repo: Arc<dyn RoleRepository>,
}

impl ListRolesUseCaseImpl {
    pub fn new(role_repo: Arc<dyn RoleRepository>) -> Self {
        ListRolesUseCaseImpl { role_repo }
    }
}

#[async_trait]
impl ListRolesUseCase for ListRolesUseCaseImpl {
    async fn execute(&self, input: ListRolesInput) -> Result<ListRolesOutput, AppError> {
        let page = if input.page < 1 { 1 } else { input.page };
        let page_size = if input.page_size < 1 {
            20
        } else if input.page_size > 100 {
            100
        } else {
            input.page_size
        };

        let offset = ((page - 1) * page_size) as usize;
        let (roles, total) = self.role_repo.find_all(offset, page_size as usize).await?;

        Ok(ListRolesOutput {
            roles,
            total,
            page,
            page_size,
        })
    }
}
